File: meresco/lucene/composedquery.py
## begin license ##
#
# "Meresco Lucene" is a set of components and tools to integrate Lucene (based on PyLucene) into Meresco
#
# Copyright (C) 2013-2016 Seecr (Seek You Too B.V.) http://seecr.nl
# Copyright (C) 2013-2014 Stichting Bibliotheek.nl (BNL) http://www.bibliotheek.nl
# Copyright (C) 2015-2016 Koninklijke Bibliotheek (KB) http://www.kb.nl
# Copyright (C) 2016 Stichting Kennisnet http://www.kennisnet.nl
#
# This file is part of "Meresco Lucene"
#
# "Meresco Lucene" is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# "Meresco Lucene" is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with "Meresco Lucene"; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
## end license ##

from .utils import simplifiedDict


class ComposedQuery(object):
    def __init__(self, resultsFromCore, query=None):
        self.cores = set()
        self._queries = {}
        self._filterQueries = {}
        self._facets = {}
        self._drilldownQueries = {}
        self._otherCoreFacetFilters = {}
        self._rankQueries = {}
        self._matches = {}
        self._unites = []
        self._sortKeys = []
        self.resultsFrom = resultsFromCore
        if query:
            self.setCoreQuery(resultsFromCore, query=query)
        else:
            self.cores.add(resultsFromCore)


    def _makeProperty(name, defaultValue=None):
        return property(
            fget=lambda self: getattr(self, name, defaultValue),
            fset=lambda self, value: setattr(self, name, value)
        )

    stop = _makeProperty('_stop')
    start = _makeProperty('_start')
    sortKeys = _makeProperty('_sortKeys')
    suggestionRequest = _makeProperty('_suggestionRequest')
    dedupField = _makeProperty('_dedupField')
    dedupSortField = _makeProperty('_dedupSortField')
    storedFields = _makeProperty('_storedFields')
    clustering = _makeProperty('_clustering')
    clusteringConfig = _makeProperty('_clusteringConfig')
    unqualifiedTermFields = _makeProperty('_unqualifiedTermFields')
    rankQueryScoreRatio = _makeProperty('_rankQueryScoreRatio')

    del _makeProperty


    def setCoreQuery(self, core, query, filterQueries=None, facets=None):
        self.cores.add(core)
        self._queries[core] = query
        if not filterQueries is None:
            for filterQuery in filterQueries:
                self.addFilterQuery(core, filterQuery)
        if not facets is None:
            for facet in facets:
                self.addFacet(core, facet)
        return self

    def addFilterQuery(self, core, query):
        self.cores.add(core)
        self._filterQueries.setdefault(core, []).append(query)
        return self

    def addFacet(self, core, facet):
        self.cores.add(core)
        self._facets.setdefault(core, []).append(facet)
        return self

    def addDrilldownQuery(self, core, drilldownQuery):
        self.cores.add(core)
        self._drilldownQueries.setdefault(core, []).append(drilldownQuery)
        return self

    def addOtherCoreFacetFilter(self, core, query):
        self.cores.add(core)
        self._otherCoreFacetFilters.setdefault(core, []).append(query)
        return self

    def setRankQuery(self, core, query):
        self.cores.add(core)
        self._rankQueries[core] = query
        return self

    def addMatch(self, matchCoreASpec, matchCoreBSpec):
        self._matches[(matchCoreASpec['core'], matchCoreBSpec['core'])] = (matchCoreASpec, matchCoreBSpec)
        resultsFromCoreSpecFound = False
        for matchCoreSpec in [matchCoreASpec, matchCoreBSpec]:
            coreName = matchCoreSpec['core']
            if coreName == self.resultsFrom:
                resultsFromCoreSpecFound = True
                try:
                    matchCoreSpec['uniqueKey']
                except KeyError:
                    raise ValueError("Match for result core '%s' must have a uniqueKey specification." % self.resultsFrom)
        if not resultsFromCoreSpecFound:
            raise ValueError("Match that does not include resultsFromCore ('%s') not yet supported" % self.resultsFrom)
        return self

    def addUnite(self, uniteCoreASpec, uniteCoreBSpec):
        if len(self.unites) > 0:
            raise ValueError("No more than 1 addUnite supported")
        for uniteCoreSpec in (uniteCoreASpec, uniteCoreBSpec):
            self.cores.add(uniteCoreSpec['core'])
        self._unites.append(Unite(self, uniteCoreASpec, uniteCoreBSpec))
        return self

    def addSortKey(self, sortKey):
        core = sortKey.get('core', self.resultsFrom)
        self.cores.add(core)
        self._sortKeys.append(sortKey)

    def queryFor(self, core):
        return self._queries.get(core)

    def filterQueriesFor(self, core):
        return self._filterQueries.get(core, [])

    def facetsFor(self, core):
        return self._facets.get(core, [])

    def drilldownQueriesFor(self, core):
        return self._drilldownQueries.get(core, [])

    def otherCoreFacetFiltersFor(self, core):
        return self._otherCoreFacetFilters.get(core, [])

    def rankQueryFor(self, core):
        return self._rankQueries.get(core)

    def keyName(self, core, otherCore):
        if core == otherCore: #TODO: Needed for filters/rank's in same core as queried core
            for matchCoreASpec, matchCoreBSpec in self._matches.values():
                if matchCoreASpec['core'] == core:
                    coreSpec = matchCoreASpec
                    break
                elif matchCoreBSpec['core'] == core:
                    coreSpec = matchCoreBSpec
                    break
        else:
            coreSpec, _ = self._matchCoreSpecs(core, otherCore)
        return coreSpec.get('uniqueKey', coreSpec.get('key'))

    def keyNames(self, core):
        keyNames = set()
        for coreName in self.cores:
            if coreName != core:
                keyNames.add(self.keyName(core, coreName))
        return keyNames

    def queriesFor(self, core):
        return [q for q in [self.queryFor(core)] + self.filterQueriesFor(core) if q]

    @property
    def unites(self):
        return self._unites[:]

    @property
    def filterQueries(self):
        return self._filterQueries.items()

    @property
    def numberOfUsedCores(self):
        return len(self.cores)

    def isSingleCoreQuery(self):
        return self.numberOfUsedCores == 1

    def coresInMatches(self):
        return set(c for matchKey in self._matches.keys() for c in matchKey)

    def validate(self):
        for core in self.cores:
            if core == self.resultsFrom:
                continue
            try:
                self._matchCoreSpecs(self.resultsFrom, core)
            except KeyError:
                raise ValueError("No match set for cores %s" % str((self.resultsFrom, core)))

    def convertWith(self, **converts):
        def convertQuery(core, query):
            if query is None:
                return None
            convertFunction = converts[core]
            if core == self.resultsFrom:
                kwargs = {'composedQuery': self}
                if self.unqualifiedTermFields:
                    kwargs['unqualifiedTermFields'] = self.unqualifiedTermFields
                return convertFunction(query, **kwargs)
            return convertFunction(query)
        self._queries = dict((core, convertQuery(core, v)) for core, v in self._queries.items())
        self._filterQueries = dict((core, [convertQuery(core, v) for v in values]) for core, values in self._filterQueries.items())
        self._rankQueries = dict((core, convertQuery(core, v)) for core, v in self._rankQueries.items())
        for unite in self._unites:
            unite.convertQuery(convertQuery)
        self._otherCoreFacetFilters = dict((core, [convertQuery(core, v) for v in values]) for core, values in self._otherCoreFacetFilters.items())

    def asDict(self):
        result = dict(vars(self))
        result['_matches'] = dict(('->'.join(key), value) for key, value in result['_matches'].items())
        result['_unites'] = [unite.asDict() for unite in self._unites]
        result['cores'] = list(self.cores)
        return result

    @classmethod
    def fromDict(cls, dct):
        cq = cls(dct['resultsFrom'])
        matches = dct['_matches']
        dct['_matches'] = dict((tuple(key.split('->')), value) for key, value in matches.items())
        dct['_unites'] = [Unite.fromDict(cq, uniteDict) for uniteDict in dct['_unites']]
        dct['cores'] = set(dct['cores'])
        for attr, value in dct.items():
            setattr(cq, attr, value)
        return cq

    def _matchCoreSpecs(self, *cores):
        try:
            coreASpec, coreBSpec = self._matches[cores]
        except KeyError:
            coreBSpec, coreASpec = self._matches[tuple(reversed(cores))]
        return coreASpec, coreBSpec

    def __repr__(self):
        return "%s%s" % (self.__class__.__name__, self.asDict())

    def infoDict(self):
        return {
            'type': self.__class__.__name__,
            'query': simplifiedDict(dict((k.replace('_', ''), v) for k,v in self.asDict().items()))
        }


class Unite(object):
    def __init__(self, parent, coreASpec, coreBSpec):
        self._parent = parent
        self.coreASpec = coreASpec
        self.coreBSpec = coreBSpec

    def queries(self):
        keyNameA = self._parent.keyName(self.coreASpec['core'], self.coreBSpec['core'])
        keyNameB = self._parent.keyName(self.coreBSpec['core'], self.coreASpec['core'])
        resultKeyName = keyNameA if self._parent.resultsFrom == self.coreASpec['core'] else keyNameB
        yield dict(core=self.coreASpec['core'], query=self.coreASpec['query'], keyName=keyNameA), resultKeyName
        yield dict(core=self.coreBSpec['core'], query=self.coreBSpec['query'], keyName=keyNameB), resultKeyName

    def convertQuery(self, convertQueryFunction):
        for spec in [self.coreASpec, self.coreBSpec]:
            spec['query'] = convertQueryFunction(spec['core'], spec['query'])

    def asDict(self):
        return {'A': [self.coreASpec['core'], self.coreASpec['query']], 'B': [self.coreBSpec['core'], self.coreBSpec['query']]}

    @classmethod
    def fromDict(cls, parent, dct):
        return cls(parent, dict(core=dct['A'][0], query=dct['A'][1]), dict(core=dct['B'][0], query=dct['B'][1]))

File: src/org/meresco/lucene/ComposedQuery.java

/* begin license *
 *
 * "Meresco Lucene" is a set of components and tools to integrate Lucene (based on PyLucene) into Meresco
 *
 * Copyright (C) 2015-2016 Koninklijke Bibliotheek (KB) http://www.kb.nl
 * Copyright (C) 2015-2016, 2018-2020 Seecr (Seek You Too B.V.) https://seecr.nl
 * Copyright (C) 2016, 2020 Stichting Kennisnet https://www.kennisnet.nl
 *
 * This file is part of "Meresco Lucene"
 *
 * "Meresco Lucene" is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * "Meresco Lucene" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with "Meresco Lucene"; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * end license */

package org.meresco.lucene;

import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.json.Json;
import javax.json.JsonArray;
import javax.json.JsonObject;
import javax.json.JsonString;
import javax.json.JsonValue;

import org.apache.lucene.search.Query;
import org.meresco.lucene.JsonQueryConverter.FacetRequest;


public class ComposedQuery {
    public String resultsFrom;
    public Set<String> cores = new HashSet<String>();
    private Map<String, Query> queries = new HashMap<String, Query>();
    public Query relationalFilter;
    public Map<String, List<Query>> filterQueries = new HashMap<String, List<Query>>();
    public Map<String, List<Query>> excludeFilterQueries = new HashMap<String, List<Query>>();
    private Map<String, List<FacetRequest>> facets = new HashMap<String, List<FacetRequest>>();
    private Map<String, List<String[]>> drilldownQueries = new HashMap<String, List<String[]>>();
    private Map<String, List<Query>> otherCoreFacetsFilter = new HashMap<String, List<Query>>();
    private Map<String, Query> rankQueries = new HashMap<String, Query>();
    public Float rankQueryScoreRatio;
    private List<Unite> unites = new ArrayList<Unite>();
    public List<Match> matches = new ArrayList<Match>();
    public QueryData queryData = new QueryData();
    public ClusterConfig clusterConfig;

    public String toString() {
        return "ComposedQuery(resultsFrom=" + resultsFrom + ",\n    " +
            "cores=" + cores + ",\n    " +
            "queries=" + queries + ",\n    " +
            "relationalFilter=" + relationalFilter + ",\n    " +
            "filterQueries=" + filterQueries + ",\n    " +
            "excludeFilterQueries=" + excludeFilterQueries + ",\n    " +
            "facets=" + facets + ",\n    " +
            "drilldownQueries=" + drilldownQueries + ",\n    " +
            "otherCoreFacetsFilter=" + otherCoreFacetsFilter + ",\n    " +
            "rankQueries=" + rankQueries + ",\n    " +
            "rankQueryScoreRatio=" + rankQueryScoreRatio + ",\n    " +
            "unites=" + unites + ",\n    " +
            "matches=" + matches + ",\n    " +
            "queryData=" + queryData + ",\n    " +
            "clusterConfig=" + clusterConfig + ")";
    }

    public ComposedQuery(String resultsFrom) {
        this.resultsFrom = resultsFrom;
    }

    public ComposedQuery(String resultsFrom, Query query) {
        this.resultsFrom = resultsFrom;
        this.queries.put(resultsFrom, query);
    }

    public static ComposedQuery fromJsonString(Reader jsonStringReader, Map<String, JsonQueryConverter> queryConverters) {
        if (jsonStringReader == null) {
            return null;
        }
        JsonObject json = Json.createReader(jsonStringReader).readObject();
        String resultsFrom = json.getString("resultsFrom");
        ComposedQuery cq = new ComposedQuery(resultsFrom);
        if (json.containsKey("_start") && json.get("_start") != JsonValue.NULL) {
            cq.queryData.start = json.getInt("_start");
        }
        if (json.containsKey("_stop") && json.get("_stop") != JsonValue.NULL) {
            cq.queryData.stop = json.getInt("_stop");
        }
        JsonQueryConverter queryConverter = queryConverters.get(resultsFrom);
        if (json.containsKey("_suggestionRequest") && json.get("_suggestionRequest") != JsonValue.NULL) {
            cq.queryData.suggestionRequest = queryConverter.convertToSuggestionRequest(json.getJsonObject("_suggestionRequest"));
        }
        cq.queryData.sort = queryConverter.convertToSort(json.getJsonArray("_sortKeys"));
        cq.queryData.dedupField = json.getString("_dedupField", null);
        cq.queryData.getDedupSortFieldsFromJson(json, "_dedupSortField");
        cq.queryData.clustering = json.getBoolean("_clustering", false);
        JsonArray fields = json.getJsonArray("_storedFields");
        if (fields != null) {
            fields.stream().forEach(s -> cq.queryData.storedFields.add(((JsonString) s).getString()));
        }
        if (json.containsKey("_clusteringConfig")) {
        	cq.queryData.clusterConfig = ClusterConfig.parseFromJsonObject(json.getJsonObject("_clusteringConfig"));
        }
        if (json.containsKey("cores")) {
            JsonArray jsonCores = json.getJsonArray("cores");
            for (int i = 0; i < jsonCores.size(); i++) {
                cq.cores.add(jsonCores.getString(i));
            }
        }
        if (json.containsKey("_queries")) {
            JsonObject queries = json.getJsonObject("_queries");
            for (String core : queries.keySet()) {
                cq.setCoreQuery(core, queryConverters.get(core).convertToQuery(queries.getJsonObject(core)));
            }
        }
        if (json.containsKey("_relationalFilterJson")) {
            String relationalFilterJson = json.getString("_relationalFilterJson");
            StringReader reader = new StringReader(relationalFilterJson);
            JsonObject relationalFilterJsonObject = Json.createReader(reader).readObject();
            reader.close();
            cq.relationalFilter = queryConverters.get(resultsFrom).convertToQuery(relationalFilterJsonObject);
        }
        if (json.containsKey("_filterQueries")) {
            JsonObject filterQueries = json.getJsonObject("_filterQueries");
            for (String coreName : filterQueries.keySet()) {
                JsonArray queries = filterQueries.getJsonArray(coreName);
                for (int i = 0; i < queries.size(); i++) {
                    cq.addFilterQuery(coreName, queryConverters.get(coreName).convertToQuery(queries.getJsonObject(i)));
                }
            }
        }
        if (json.containsKey("_excludeFilterQueries")) {
            JsonObject excludeFilterQueries = json.getJsonObject("_excludeFilterQueries");
            for (String coreName : excludeFilterQueries.keySet()) {
                JsonArray queries = excludeFilterQueries.getJsonArray(coreName);
                for (int i = 0; i < queries.size(); i++) {
                    cq.addExcludeFilterQuery(coreName, queryConverters.get(coreName).convertToQuery(queries.getJsonObject(i)));
                }
            }
        }
        if (json.containsKey("_otherCoreFacetFilters")) {
            JsonObject filterQueries = json.getJsonObject("_otherCoreFacetFilters");
            for (String coreName : filterQueries.keySet()) {
                JsonArray queries = filterQueries.getJsonArray(coreName);
                for (int i = 0; i < queries.size(); i++) {
                    cq.addOtherCoreFacetFilter(coreName, queryConverters.get(coreName).convertToQuery(queries.getJsonObject(i)));
                }
            }
        }
        if (json.containsKey("_drilldownQueries")) {
            JsonObject drilldownQueries = json.getJsonObject("_drilldownQueries");
            for (String coreName : drilldownQueries.keySet()) {
                JsonArray ddQueries = drilldownQueries.getJsonArray(coreName);
                for (int i = 0; i < ddQueries.size(); i++) {
                    JsonArray ddQuery = ddQueries.getJsonArray(i);
                    JsonArray jsonPath = ddQuery.getJsonArray(1);
                    String[] path = new String[jsonPath.size()];
                    for (int j=0; j < jsonPath.size(); j++) {
                        path[j] = jsonPath.getString(j);
                    }
                    cq.addDrilldownQuery(coreName, ddQuery.getString(0), path);
                }
            }
        }
        if (json.containsKey("_facets")) {
            JsonObject jsonFacets = json.getJsonObject("_facets");
            for (String coreName : jsonFacets.keySet()) {
                for (FacetRequest facetRequest : queryConverters.get(coreName).convertToFacets(jsonFacets.getJsonArray(coreName))) {
                    cq.addFacet(coreName, facetRequest);
                }
            }
        }
        if (json.containsKey("_matches")) {
            JsonObject matches = json.getJsonObject("_matches");
            for (String match : matches.keySet()) {
                String[] coreNames = match.split("->");
                JsonArray coreDicts = matches.getJsonArray(match);
                JsonObject coreSpec1 = coreDicts.getJsonObject(0);
                JsonObject coreSpec2 = coreDicts.getJsonObject(1);
                String keyName1 = coreSpec1.getString("key", coreSpec1.getString("uniqueKey", null));
                String keyName2 = coreSpec2.getString("key", coreSpec2.getString("uniqueKey", null));
                if (coreSpec1.getString("core").equals(coreNames[0])) {
                    cq.addMatch(coreNames[0], coreNames[1], keyName1, keyName2);
                } else {
                    cq.addMatch(coreNames[0], coreNames[1], keyName2, keyName1);
                }
            }
        }
        if (json.containsKey("_rankQueries")) {
            JsonObject rankQueries = json.getJsonObject("_rankQueries");
            for (String coreName : rankQueries.keySet()) {
                cq.setRankQuery(coreName, queryConverters.get(coreName).convertToQuery(rankQueries.getJsonObject(coreName)));
            }
        }
        if (json.containsKey("_rankQueryScoreRatio")) {
            cq.rankQueryScoreRatio = (float) json.getJsonNumber("_rankQueryScoreRatio").doubleValue();
        }
        if (json.containsKey("_unites")) {
            JsonArray unites = json.getJsonArray("_unites");
            for (int i = 0; i < unites.size(); i++) {
                JsonArray A = unites.getJsonObject(i).getJsonArray("A");
                JsonArray B = unites.getJsonObject(i).getJsonArray("B");
                String core1 = A.getString(0);
                Query query1 = queryConverters.get(core1).convertToQuery(A.getJsonObject(1));
                String core2 = B.getString(0);
                Query query2 = queryConverters.get(core2).convertToQuery(B.getJsonObject(1));
                cq.addUnite(core1, query1, core2, query2);
            }
        }
        return cq;
    }

    public void addMatch(String core1, String core2, String core1Key, String core2Key) {
        matches.add(new Match(core1, core2, core1Key));
        matches.add(new Match(core2, core1, core2Key));
        cores.add(core1);
        cores.add(core2);
    }

    public Query queryFor(String core) {
        return this.queries.get(core);
    }

    public List<Query> queriesFor(String core) {
        List<Query> qs = new ArrayList<Query>();
        Query otherCoreQuery = queryFor(core);
        if (otherCoreQuery != null)
            qs.add(otherCoreQuery);
        List<Query> otherCoreFilterQueries = filterQueries.get(core);
        if (otherCoreFilterQueries != null)
            qs.addAll(otherCoreFilterQueries);
        return qs;
    }

    public List<FacetRequest> facetsFor(String core) {
        if (!this.facets.containsKey(core))
            return new ArrayList<FacetRequest>();
        return this.facets.get(core);
    }

    public List<String[]> drilldownQueriesFor(String core) {
        return this.drilldownQueries.get(core);
    }

    public List<Query> otherCoreFacetFiltersFor(String core) {
        if (!this.otherCoreFacetsFilter.containsKey(core))
            return new ArrayList<Query>();
        return this.otherCoreFacetsFilter.get(core);
    }

    public Query rankQueryFor(String core) {
        return this.rankQueries.get(core);
    }

    public List<Unite> getUnites() {
        return this.unites;
    }

    public String keyName(String coreName, String otherCoreName) {
        for (Match match : matches) {
            if (match.core1.equals(coreName) && match.core2.equals(otherCoreName))
                return match.keyName;
            else if (coreName.equals(otherCoreName) && match.core1.equals(coreName))
                return match.keyName;
        }
        return null;
    }

    public String[] keyNames(String core) {
        Set<String> keyNames = new HashSet<String>();
        for (String coreName : this.cores) {
            if (!coreName.equals(core)) {
                keyNames.add(keyName(core, coreName));
            }
        }
        return keyNames.toArray(new String[0]);
    }

    public void setCoreQuery(String coreName, Query query) {
        this.queries.put(coreName, query);
    }

    public void addFilterQuery(String coreName, Query query) {
        if (filterQueries.containsKey(coreName))
            filterQueries.get(coreName).add(query);
        else {
            List<Query> queries = new ArrayList<Query>();
            queries.add(query);
            filterQueries.put(coreName, queries);
        }
    }

    public void addExcludeFilterQuery(String coreName, Query query) {
        if (excludeFilterQueries.containsKey(coreName))
            excludeFilterQueries.get(coreName).add(query);
        else {
            List<Query> queries = new ArrayList<Query>();
            queries.add(query);
            excludeFilterQueries.put(coreName, queries);
        }
    }

    public void addDrilldownQuery(String coreName, String dim, String... value) {
        if (drilldownQueries.containsKey(coreName)) {
            drilldownQueries.get(coreName).add(new String[] { dim });
            drilldownQueries.get(coreName).add(value);
        } else {
            List<String[]> fieldValuesMap = new ArrayList<String[]>();
            fieldValuesMap.add(new String[] { dim });
            fieldValuesMap.add(value);
            drilldownQueries.put(coreName, fieldValuesMap);
        }
    }

    public void addOtherCoreFacetFilter(String coreName, Query query) {
        if (this.otherCoreFacetsFilter.containsKey(coreName))
            this.otherCoreFacetsFilter.get(coreName).add(query);
        else {
            List<Query> otherCoreFacetsFilter = new ArrayList<Query>();
            otherCoreFacetsFilter.add(query);
            this.otherCoreFacetsFilter.put(coreName, otherCoreFacetsFilter);
        }
    }

    public void addUnite(String core1, Query query1, String core2, Query query2) {
        this.unites.add(new Unite(core1, core2, query1, query2));
    }

    public void addFacet(String coreName, FacetRequest facetRequest) {
        if (this.facets.containsKey(coreName))
            this.facets.get(coreName).add(facetRequest);
        else {
            List<FacetRequest> facets = new ArrayList<FacetRequest>();
            facets.add(facetRequest);
            this.facets.put(coreName, facets);
        }
    }

    public void setRankQuery(String core, Query query) {
        this.rankQueries.put(core, query);
    }

    static class Match {
        private String core1;
        private String core2;
        private String keyName;

        public String toString() {
            return "Match(" + core1 + ", " + core2 + ", " + keyName + ")";
        }

        public Match(String core1, String core2, String keyName) {
            this.core1 = core1;
            this.core2 = core2;
            this.keyName = keyName;
        }
    }

    static class Unite {
        public String coreA;
        public String coreB;
        public Query queryA;
        public Query queryB;

        public String toString() {
            return "Unite(" + coreA + ", " + coreB + ", " + queryA + ", " + queryB + ")";
        }

        public Unite(String core1, String core2, Query query1, Query query2) {
            this.coreA = core1;
            this.coreB = core2;
            this.queryA = query1;
            this.queryB = query2;
        }
    }
}
